### PR TITLE
Implements smoothing and parallelisation

### PR DESCRIPTION
**Pull request by jleechung**
This PR implements label smoothing and parallel clustering. 

Other changes include:
- Filtering out samples with NAs in their features on init.
- Seed setting for clustering
- Version dependency on leidenAlg (>1.1.0) for compatibility with igraph 
- Neighborhood-sampling (see arguments `sample_size`, `sample_renorm`, and `seed` for function `ComputeBanksy`)
- Replacing `M` with `compute_agf` and `use_agf` as arguments for specifying which harmonics to compute

This will be the last release / update for BANKSY with BanksyObject. Next release will adopt SingleCellExperiment as a container.

### CODE REVIEW

File: R/computation.R [hunk 1]
```diff
@@ -755,3 +768,8 @@ subsampler = function(knnDF,
     if (sample_renorm) x[, weight := weight / sum(weight), by = from]
     data.table(x)
 }
+
```
Comment (jleechung) [inline R/computation.R:771]: For controlling harmonic usage, replace `M` with `compute_agf` (and `use_agf`).
This function converts the boolean flags `compute_agf` and `use_agf` to the corresponding `M` value. 
`M` can still be used in a backward-compatible manner in `ComputeBanksy`, `RunBanksyPCA`, `RunBanksyUMAP` and `ClusterBanksy`; if it is provided, it overwrites `compute_agf` (or `use_agf`). This is documented as advanced usage.

Comment (vipulsinghal02) [reply R/computation.R:771]: Sounds good.

File: R/clustering.R [hunk 2]
```diff
@@ -436,3 +520,123 @@ getARI <- function(bank, digits = 3) {
...
+                        verbose = TRUE,
+                        include_self = TRUE) {
+    
+    if (is.null(cluster_names)) {
+        cluster_names = clust.names(bank)
+        if (is.null(cluster_names)) stop('No cluster names found')
+    } 
+    if (is.list(bank@own.expr)) {
+        # Multiple datasets
+        return(bank)
```
Comment (vipulsinghal02) [inline R/clustering.R:575]: I suppose down the road we will also do smoothing for the multi dataset mode?

**Review by vipulsinghal02 — APPROVED** [review]
Overall looks good. Let's proceed with the merge.

File: R/clustering.R (hunk 2 above)
Comment (jleechung) [reply R/clustering.R:575]: yes!

**Pull request merged**